### ISSUE TITLE
We need to test before release. `linkml-runtime` release v1.6.1 is broken, and it's impacting the rest of the supply chain (e.g. `bmt` toolkit)

### ISSUE THREAD

**Issue by vemonet** [linkml/linkml#1749]
**Describe the bug**
A clear and concise description of what the bug is (including the error message that is printed, if any)

We are using `bmt` to get some ancestors. But since the release of `linkml-runtime` version `1.6.1` creating the `Toolkit()` fails due to error with wikidata URI inside `linkml-runtime`

**To reproduce**
1. `pip install "linkml-runtime ==1.6.1"`
2. Create python script with:
```python
from bmt import Toolkit

biolink = Toolkit()
```
3. Run the script to see the error

**Expected behavior**
Such a basic action should never fails in a released package. 

**Screenshots**

Here is the error trace:

```
File "/usr/local/lib/python3.10/dist-packages/trapi_predict_kit/trapi_parser.py", line 10, in <module>
    biolink = Toolkit()
  File "/usr/local/lib/python3.10/dist-packages/bmt/toolkit.py", line 52, in __init__
    self.view = SchemaView(schema)
  File "/usr/local/lib/python3.10/dist-packages/linkml_runtime/utils/schemaview.py", line 138, in __init__
    schema = load_schema_wrap(schema)
  File "/usr/local/lib/python3.10/dist-packages/linkml_runtime/utils/schemaview.py", line 70, in load_schema_wrap
    schema = yaml_loader.load(path, target_class=SchemaDefinition, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/linkml_runtime/loaders/loader_root.py", line 71, in load
    results = self.load_any(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/linkml_runtime/loaders/yaml_loader.py", line 41, in load_any
    return self._construct_target_class(data_as_dict, target_class)
  File "/usr/local/lib/python3.10/dist-packages/linkml_runtime/loaders/loader_root.py", line 132, in _construct_target_class
    return target_class(**data_as_dict)
  File "<string>", line 66, in __init__
  File "/usr/local/lib/python3.10/dist-packages/linkml_runtime/linkml_model/meta.py", line 591, in __post_init__
    self._normalize_inlined_as_dict(slot_name="slots", slot_type=SlotDefinition, key_name="name", keyed=True)
  File "/usr/local/lib/python3.10/dist-packages/linkml_runtime/utils/yamlutils.py", line 100, in _normalize_inlined_as_dict
    self._normalize_inlined(slot_name, slot_type, key_name, keyed, False)
  File "/usr/local/lib/python3.10/dist-packages/linkml_runtime/utils/yamlutils.py", line 205, in _normalize_inlined
    form_1({k: v})
  File "/usr/local/lib/python3.10/dist-packages/linkml_runtime/utils/yamlutils.py", line 153, in form_1
    order_up(key, slot_type(**as_dict(raw_obj)))
  File "<string>", line 115, in __init__
  File "/usr/local/lib/python3.10/dist-packages/linkml_runtime/linkml_model/meta.py", line 2320, in __post_init__
    super().__post_init__(**kwargs)
  File "/usr/local/lib/python3.10/dist-packages/linkml_runtime/linkml_model/meta.py", line 862, in __post_init__
    super().__post_init__(**kwargs)
  File "/usr/local/lib/python3.10/dist-packages/linkml_runtime/linkml_model/meta.py", line 460, in __post_init__
    self.exact_mappings = [v if isinstance(v, URIorCURIE) else URIorCURIE(v) for v in self.exact_mappings]
  File "/usr/local/lib/python3.10/dist-packages/linkml_runtime/linkml_model/meta.py", line 460, in <listcomp>
    self.exact_mappings = [v if isinstance(v, URIorCURIE) else URIorCURIE(v) for v in self.exact_mappings]
  File "/usr/local/lib/python3.10/dist-packages/linkml_runtime/utils/metamodelcore.py", line 104, in __init__
    raise ValueError(f"{v} is not a valid URI or CURIE")
ValueError: WIKIDATA_PROPERTY:P854 is not a valid URI or CURIE
```

**About your computer (if applicable, please complete the following information):**
 - OS: linux 
 - Browser not relevant 
 - Version `linkml-runtime  1.6.1`

**Additional context**

There should be tests in place that **prevents people from releasing a new version if the tests are not passing**. It is not normal that we manage to release packages that are fundamentally broken, it breaks all of the supply chain in way that can be hard to debug!

In this specific case there should be a step that runs the tests before publishing the new release to pypi: https://github.com/linkml/linkml-runtime/blob/main/.github/workflows/pypi-publish.yaml 

If we check globally the latest runs of the test workflow we can see it is constantly breaking: https://github.com/linkml/linkml-runtime/actions 

I think it is really important that **every single** that is part of the Translator supply chain should follow those practices:
* Every packages should have basic overall integration tests that checks if the main features are working (no need for every edge cases, at least the main features)
* Those tests needs to be run automatically at each push
* Those tests needs to be run automatically at each release, and the release should be cancelled if the tests don't pass

I know a lot of people think it's a chore, but automated testing is really paramount, especially in such a decentralized system like ours

Let me know, I can take a bit of time to go other every Translator projects to improve those practices and make sure **proper testing is done before releasing**. If we are waiting for everyone to do it themselves we will still continue to have inconsistencies for years :)

**Comment by rly** [comment]
If I may add my two cents... the NWB software ecosystem has a similar interlinked set of repositories as LinkML. To help catch upstream changes that break something downstream, we have github actions workflows that run on each PR that check out a downstream repo, install the PR branch of the upstream repo (linkml-runtime here), and run the basic tests of the downstream repo (bmt here?). e.g., https://github.com/hdmf-dev/hdmf/blob/dev/.github/workflows/run_pynwb_tests.yml . I do not see a similar workflow here but maybe I am not looking in the right places. This could get expensive in terms of actions runtime, but we have found it to be very useful in catching bugs or accidental breaking changes.

**Comment by vemonet** [comment]
> I found the bug manually as I was upgrading Biolink and BMT repos, and as part of the fix, we also https://github.com/linkml/linkml-runtime/pull/284.

Thanks! I think I find why we did not spotted it earlier. If the test suite is complete, normally we should not find the bug manually, it should make the CI tests break

In this case the bug only appeared when using the newer version of `linkml-runtime` 

The current tests workflow uses the repository [`poetry.lock` file and cache for dependencies](https://github.com/linkml/linkml/blob/main/.github/workflows/main.yaml#L71). So we can't spot this kind of issues related to the latest release versions of dependencies until someone manually update the `poetry.lock` file and commit the new lock



When we publish a python package the `poetry.lock` is not published (it is only used in development), which means that:
* We test in an environment with locked dependencies version, not necessarily the latest one
* When a user install our package, the dependency manager will always try to resolve to the latest available version for each dependency

For integration tests to be efficient, they need to be the closest possible to the experience the user will have using the package. In this case this means the tests needs to be run after resolving the latest version of each dependency (instead of using locked versions).

Otherwise you would need to manually run an update of the `poetry.lock` everytime you want to get the latest versions of each dependencies (since we have so many dependencies that means everyday)

## The solution

We should **remove the `poetry.lock` and caching from the test workflows**. Yes it will take a few more seconds to install, but at every run you will test the package against the latest version available for each dependency (and we don't care if the test workflow takes 4min instead of 3min)

Which allows to spot much faster and consistently errors related to new dependencies versions.

Then we will need to do this for every package in the Translator ecosystem!

You could also completely remove the `poetry.lock` from the repository. This will remove the need to commit updates of it. The first `poetry install` will be a bit longer, and the different devs working on the project will get a slightly different versions, but that's not a problem. The package is supposed to work the same and pass the tests for all versions of its dependencies, so if it fails for one of the devs because they have a wrong version that means they find a bug, and the dependencies needs to be updated to exclude this broken version!

`poetry.lock` is nice for an application deployment that needs to be reproducible, but for a package it is not really helpful. If we continue to use `poetry.lock` and caching then we will continue to regularly have those kind of issues where a new

**Comment by rly** [comment]
+1 to the above and the difference between libraries and applications. I would recommend what we do in the NWB repos, which are libraries like LinkML: 
- A workflow job that tests the dev-pinned dependencies (though not necessary)
- One that tests the latest allowed version of each dependency
- One that tests the latest allowed version of each dependency including pre-releases (runs daily instead of per-PR so we don't waste cycles)
- One that tests the minimum allowed version of each dependency (including Python version)

Devs are still encouraged to install the dev-pinned dependencies to make testing and reproducing issues easier.

**Comment by vemonet** [comment]
Exactly @rly, you are going even further than me with the test of "each dependency including pre-releases", but that is really the way to go to have something stable

Do you have a pointer of a repository that implements the tests on pre-releases? Just to see how it can be done easily

This is exactly what we need implemented in **every** package built for the Translator :D

**Comment by rly** [comment]
> Do you have a pointer of a repository that implements the tests on pre-releases? Just to see how it can be done easily

We are not using poetry so I do not know how it would transfer, but in a workflow, we basically install the repo using `pip install -U --pre .` and then run the tests. Easy!

**Comment by sierra-moxon** [comment]
I took your advice, @rly and added some actions downstream on `Biolink` and `bmt` that cover a couple of these suggestions:
- remove the poetry.lock file and re-install in the action, then run tests on the fresh install 
- install a downstream consumer of the package, uninstall the old version of the package, re-install the dev version and run the tests in the downstream package (in my case, this is bmt cloning and running tests on kgx, one of its dependencies).  

I think we might be able to do similar checks here, for linkml as a consumer of linkmk-runtime.  I've added the "December Developer Days" label here and will try to implement it in the next release (though I take your point about creating actions that will fail with intentional breaking changes).
